File: code_examples/4-Functions/7-Recursion/Java/Main.java
/*
 * This program calculates factorial
 *
 * @author  Mr Coxall
 * @version 1.0
 * @since   2020-09-01
 */

import java.util.Scanner;

public class Main {
    static int factorial(int number) {
      // recursive function calculating factorial
      if (number == 1) {
          return 1;
      } else {
          return number * factorial(number - 1);
      }
    }

  public static void main(String[] args) {
    // this function gets user input
    int userInputInt;

    // input
    Scanner scanner = new Scanner(System.in);
    System.out.print("Enter a positive integer: ");
    userInputInt = scanner.nextInt();

    if (userInputInt < 0) {
      System.out.println("That was not a positive integer.");
    } else {
      // call function
      int factorialAnswer = factorial(userInputInt);
      System.out.printf("The factorial of %d is %d%n", userInputInt, factorialAnswer);
    }

    System.out.println("\nDone.");
  }
}

File: code_examples/4-Functions/7-Recursion/Python/main.py
#!/usr/bin/env python3
"""
Created by: Mr. Coxall
Created on: Sep 2020
This module calculates factorial
"""


def factorial(number: int) -> int:
    """ A recursive function that returns the factorial of a number"""
    if number == 1:
        return 1
    else:
        return number * factorial(number - 1)


def main():
    """this function gets user input"""

    # input
    user_input_int = int(input("Enter a positive integer: "))

    if user_input_int < 0:
        print("That was not a positive integer.")
    else:
        # call function
        factorial_answer = factorial(user_input_int)
        print(f"The factorial of {user_input_int} is {factorial_answer}")

    print("\nDone.")


if __name__ == "__main__":
    main()
